toolhelp/toolhelp.c

```diff
@@ -534,21 +541,101 @@ BOOL16 WINAPI StackTraceNext16(STACKTRACEENTRY *ste)
     return TRUE;
 }
 
+static int lastcb = 0;
+static SEGPTR lastaddr = 0;
+
+void WINAPI next_intcb(CONTEXT *context)
+{
+    WORD *stkptr = (WORD *)MapSL(MAKESEGPTR(context->SegSs, LOWORD(context->Esp)));
+    if (lastcb == nrofintcbs)
+    {
+        lastcb = 0;
+        context->Esp += 12;
+        context->Eip = stkptr[3];
+        context->SegCs = stkptr[4];
+        context->ContextFlags = stkptr[5];
+        lastaddr = MAKESEGPTR(context->SegCs, context->Eip);
+        return;
+    }
+    stkptr -= 2;
+    FARPROC16 next_intcb16 = GetProcAddress16(GetModuleHandle16("TOOLHELP"), "next_intcb");
+    stkptr[0] = OFFSETOF(next_intcb16);
+    stkptr[1] = SELECTOROF(next_intcb16);
+    context->Esp -= 4;
+    lastcb++;
+    context->Eip = OFFSETOF(intcbs[lastcb].lpfnIntCallback);
+    context->SegCs = SELECTOROF(intcbs[lastcb].lpfnIntCallback);
+}
+
+FARPROC16 WINAPI get_intcb(SEGPTR *stack, SEGPTR addr, WORD flags, WORD err, WORD ax)
+{
+    if(!nrofintcbs)
+        return FALSE;
+    if(addr == lastaddr) // failed to handle once
+        return FALSE;
+    lastaddr = 0;
+
+    WORD *stkptr = (WORD *)MapSL(*stack);
+    FARPROC16 next_intcb16 = GetProcAddress16(GetModuleHandle16("TOOLHELP"), "next_intcb");
+    stkptr -= 8;
+    stkptr[0] = OFFSETOF(next_intcb16);
+    stkptr[1] = SELECTOROF(next_intcb16);
+    stkptr[2] = ax;
+    stkptr[3] = err;
+    stkptr[4] = 0;
+    stkptr[5] = OFFSETOF(addr);
+    stkptr[6] = SELECTOROF(addr);
+    stkptr[7] = flags;
+    *stack -= 16;
+    lastcb = 1;
+    return intcbs[0].lpfnIntCallback;
+}        
+
 /***********************************************************************
  *		InterruptRegister (TOOLHELP.75)
  */
-BOOL16 WINAPI InterruptRegister16( HTASK16 task, FARPROC callback )
+BOOL16 WINAPI InterruptRegister16( HTASK16 htask, FARPROC16 callback )
 {
-    FIXME("(%04x, %p), stub.\n", task, callback);
+    int	i;
+
+    TRACE("(%x,%x,%x).\n", htask, (DWORD)callback);
```

Comment:
printf format takes 3 arguments, but it takes only 2 arguments.
```suggestion
    FIXME("(%04x, %04x:%04x), semi-stub.\n", task, SELECTOROF(callback), OFFSETOF(callback));
```